TrilinosWrappers::SolverDirect: adjust interface

peterrum opened the pull request:
... to be able to switch between `TrilinosWrappers::PreconditionILU`, `TrilinosWrappers::PreconditionAMG`, `TrilinosWrappers::SolverDirect`.

The changes are:
- derive from `Subscriptor`
- allow to pass `initialize()` an `AdditionalData` object
- introduce `vmult` functions
- make `SolverControl` optional

tjhei commented on include/deal.II/lac/trilinos_solver.h:
```diff
@@ -522,6 +522,11 @@ namespace TrilinosWrappers
+    /**
+     * Constructor. Creates the solver without solver control object.
+     */
+    SolverDirect(const AdditionalData &data = AdditionalData());
```
```suggestion
    explicit SolverDirect(const AdditionalData &data = AdditionalData());
```

tjhei commented on include/deal.II/lac/trilinos_solver.h:
```diff
@@ -542,6 +547,15 @@ namespace TrilinosWrappers
+     * factorization for it with the package chosen from the additional
+     * data structure. Note that there is no need for a preconditioner
+     * here and solve() is not called.
+     */
```
@p data replaces the data stored in this instance.

tjhei commented on include/deal.II/lac/trilinos_solver.h:
```diff
@@ -559,6 +573,23 @@ namespace TrilinosWrappers
 
+    /**
+     * Solve the linear system <tt>Ax=b</tt> based on the
+     * package set in initialize(). Note the matrix is not refactorized during
```
or in the constructor, right?

peterrum commented:
@tjhei I have addressed the comments.

> Are you deriving from Subscriptor so that you have a base class that you can do a dynamic_cast from? That seems like a questionable thing to do given that the interfaces of the various classes are incompatible. Typically, one would have an interface class where derived classes implement the interface. Would that work?
If that doesn't work, why don't you store these objects via std::variant? This avoids the dynamic_cast from an otherwise rather arbitrary base class.

@bangerth No. The multigrid classes store `SmartPointer` instances which require that the classes inherit from `Subscriptor`.

kronbichler commented on include/deal.II/lac/trilinos_solver.h:
```diff
@@ -559,6 +574,23 @@ namespace TrilinosWrappers
+    /**
+     * Solve the linear system <tt>Ax=b</tt> based on the
+     * package set in the constructor or initialize(). Note the matrix is not
+     * refactorized during this call.
```
```suggestion
     * refactored during this call.
```

peterrum replied:
I applied this change to the old functions as well :+1:

kronbichler reviewed (approved):
Looks good to me apart from a minor typo.
